query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java

```diff
@@ -145,77 +145,79 @@ private Object handleDataWriteCommand(InvocationContext ctx, DataWriteCommand co
       if (command.hasAnyFlag(FlagBitSets.SKIP_INDEXING)) {
          return invokeNext(ctx, command);
       }
-      CacheEntry entry = ctx.lookupEntry(command.getKey());
-      if (ctx.isInTxScope()) {
-         // replay of modifications on remote node by EntryWrappingVisitor
-         if (entry != null && !entry.isChanged() && (entry.getValue() != null || !unreliablePreviousValue(command))) {
-            Map<Object, Object> oldValues = registerOldValues((TxInvocationContext) ctx);
-            oldValues.putIfAbsent(command.getKey(), entry.getValue());
+      return invokeNextThenApply(ctx, command, (rCtx, cmd, rv) -> {
```

Comment:
lambda here and in `handleManyWriteCommand()` are "the same". you could cache it and use it in both places.

Reply:
They are not quite the same: one uses single values the other multiple values, one use WriteCommand and the other DataWriteCommand, hoe to you suggest to reuse it?

Reply:
DataWriteCommand is a subclass of WriteCommand. 
Creating a method like `updateIndexes(InvocationContext ctx, WriteCommand cmd, Object rv)` and you can use it in `invokeNextThenApply()` method.

Reply:
Not all WriteCommand are DataCommads and expose ```getKey()```, so I'd need to treat all commands with multiple keys command and put some conditions to avoid extra work, I think I gave up after 5min into the refactor :)

Reply:
ok. I had something like this in mind: https://github.com/pruivo/infinispan/commit/627bb426b60ce8e2a11a2b00f1153c2a27d3dba8

Reply:
@pruivo could you please send a follow up PR, after this one is merged?